### ISSUE TITLE
DlBlurMaskFilter implements a respect_ctm flag that seems dubious

### ISSUE THREAD

**Issue by flar** [flutter/flutter#115112]
When modifying the text package to render text directly to a DisplayList, it was discovered that the text package was constructing the Skia blur filters with a flag instructing them to ignore the CTM when applying the blur. This was apparently done to get the output to match some Material design standards, but we don't have the source material on hand to verify the need for the flag.

The problem is that this flag, while it might cause the shadows to look right in a default application with respect to some spec, will cause these shadows to not scale either due to the display DPI, or due to a rendering scaling affect applied by the application - such as a zoom feature or a scaling transition. This will violate the expectations of how the transform should affect rendering. It's possible that the shadow feature is not supposed to vary with the size of the font, or that it will not vary due to some other use of transforms during text rendering, but it probably should respond to DPI and globally applied application scaling. Since the flag disables response to any transform from any source, it is likely not the right solution for whatever design goal it was trying to match.

Further investigation needs to be done to determine why this flag was used to support shadowed text and to perhaps change the implementation to achieve the design goals a different way without the flag.

**Comment by zanderso** [comment]
From triage: This issue is about getting text shadows to look right without using this flag. Right now, Impeller is ignoring the flag, so doing this seems desirable, however it could cause visual differences. We believe this flag was set to its current value somewhat arbitrarily, and e.g. Chrome chooses the opposite value. This issue needs further investigation. The problem the flag was solving could need a different solution.

**Comment by bdero** [comment]
We just ran up against this problem in https://github.com/flutter/flutter/issues/150272. This caused a lot of confusion and we ended up implementing `respect_ctm` in order to achieve the same behavior as Skia. So I decided to spend some time researching this in hopes that we can figure out the right behavior and confidently close the loop.

But after taking a closer look, I'm convinced that the [`DisplayListParagraphPainter::drawTextShadow`](https://github.com/flutter/engine/blob/7f8033ffa7211e951ed69dc572796c44e9263fbe/third_party/txt/src/skia/paragraph_skia.cc#L122-L125) shouldn't be setting `respect_ctm` to `false` at all and we should just flip it to `true`/remove it and call it a day.

**TL;DR:** `respect_ctm` is mildly useful for implementing Material Design shadows, but **TextStyle Shadows are not actually Material Design shadows**, so they should scale up/down with the rest of the scene. Also, we probably don't need to support Material Design shadows on text either way.

## Why we should remove `respect_ctm`.

Material Design shadows are supposed to be controlled with a high level "elevation" parameter that drives the look of the shadow's penumbra based on the theoretical 3D lighting in the scene. [Impeller's DL `drawShadow`](https://github.com/flutter/engine/blob/e059e42031dc2410456b3f39c5118b13878311cd/impeller/display_list/dl_dispatcher.cc#L1144-L1148) for drawing material path shadows does this correctly, removing the CTM scale and respecting the device pixel ratio in the process of computing the correct blur for the shadow's penumbra.

So when implementing a Material Design shadow, it's kind of convenient to have a `respect_ctm` flag available alongside the device pixel ratio. You wouldn't ever _need_ one, of course. The Material path shadow implementation on the Impeller side just inverts the CTM scaling to accomplish this instead.

However, the input for `drawTextShadow` is coming directly from the framework's `TextStyle` [`Shadow`](https://api.flutter.dev/flutter/dart-ui/Shadow-class.html), which are **_not_ Material Design shadows**. They don't take an "elevation". Instead, they take a blur sigma and offset directly.

## Why we don't need to support Material Design shadows on text.

Given that `drawTextShadow` does not implement Material Design shadows, do we need to go and add support for them so that Flutter widgets can use them?

I don't think so.

From looking through the Material Design [2](https://m2.material.io/design/environment/elevation.html) and [3](https://m3.material.io/styles/elevation/overview) docs, elevation (and thus material shadows) are applicable to all "components" or "surfaces", but looking through the components lists, there don't seem to be any examples of naked text being elevated above a surface and projecting a shadow.
That's understandable; doing so would look terrible in almost every situation presented by the components list.

Putting physical-looking shadows on text seems like a recipe for illegibility, unless maybe the text is huge.
In the very rare situation where a user wants to add a physical looking shadow to their enormous WordArt text, they're probably not going to look around for a Material Design "elevation" configuration that gives them barely any control over the look of the shadow.
Instead, they're just going to play around with the TextStyle Shadow parameters without a second thought to make it look like how they want it to.

And so I think adding an interface to support Material Design shadows on text would be pretty much useless.

**Comment by flar** [comment]
Tagging @jason-simmons (who added the feature to match what had been happening with Skia) for feedback on [the last comment](https://github.com/flutter/flutter/issues/115112#issuecomment-2169127211) about getting rid of this feature.

**Comment by jason-simmons** [comment]
Setting `respectCTM = false` dates back to the original implementation of text shadows in the old Libtxt text renderer from 2018:
https://github.com/flutter/engine/commit/5f529566#diff-4d392f84bdcb936161fe95a2ace6fbe399e3b73a3b843df9060b40d351dc8fbaR1084

I don't know why the `respectCTM` flag was set to false, and I suspect that it may have been a mistake.  AFAIK other text shadow rendering implementations such as Chromium are not doing that.

I would support removing the use of `respectCTM = false`.  However, I don't know how challenging it will be for existing apps to adapt to the new behavior.

**Comment by bdero** [comment]
Yeah I find it likely that it's just accidental.

Regardless of whether or not we want text shadows to ignore the user's scale (which I don't think is the right choice), the current behavior is to also ignore the device pixel ratio, which is certainly wrong. On a low DPI monitors the shadow renders as much more spread out than on high DPI devices.

Up until now for Impeller on iOS (which has been the default renderer on iOS stable for a while now), we've been scaling all mask blurs (including text shadow mask blurs) with the CTM, like how normal mask blurs work. In [this recent patch](https://github.com/flutter/engine/pull/53377), we added support for `respect_ctm` in order to match the behavior when using Skia, which regresses iOS back to the (incorrect) behavior of disregarding the DPR/user scale.
